core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -246,18 +246,18 @@ private[spark] class MemoryStore(
         val amountToRequest = size - unrollMemoryUsedByThisBlock
         keepUnrolling = reserveUnrollMemoryForThisTask(blockId, amountToRequest, memoryMode)
         if (keepUnrolling) {
-          unrollMemoryUsedByThisBlock += amountToRequest
+          unrollMemoryUsedByThisBlock = size
         }
+      } else if (size < unrollMemoryUsedByThisBlock) {
+        releaseUnrollMemoryForThisTask(memoryMode, unrollMemoryUsedByThisBlock - size)
+        unrollMemoryUsedByThisBlock = size
       }
 
       if (keepUnrolling) {
         val entry = entryBuilder.build()
-        // Synchronize so that transfer is atomic
-        memoryManager.synchronized {
-          releaseUnrollMemoryForThisTask(memoryMode, unrollMemoryUsedByThisBlock)
-          val success = memoryManager.acquireStorageMemory(blockId, entry.size, memoryMode)
-          assert(success, "transferring unroll memory to storage memory failed")
-        }
+        // In fact, unroll memory is also storage memory, it is unnecessary to
+        // release unroll memory really
+        releaseUnrollMemoryForThisTask(memoryMode, unrollMemoryUsedByThisBlock, false)
```

Comment:
Did you check other calls on this function? Are you sure all other calls' third param default to be ```true```,  except here only?

Reply:
t am sure all other calls' third param default to be true

Reply:
Well, it seems a little awkward. How about change ```transferUnrollToStorage()```'s(which is removed after #19285, [see L242](https://github.com/apache/spark/blame/c22eaa94e85aaac649566495dcf763a5de3c8d06/core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala#L242) ) behavior , rather than ```releaseUnrollMemoryForThisTask ```?